docs/USAGE-GUIDE.md

```diff
@@ -1777,6 +1777,210 @@ handle. The handle is considered invalid after **s2n_connection_free** is used.
+Use this callback to implement custom PSK selection logic. By default s2n-tls server chooses the first identity in its PSK
+list that also appears in the client's PSK list. This callback must be implemented and set on the connection prior to the ClientHello.
+
+**s2n_config_set_psk_selection_callback** sets the callback to select the PSK to be used for the connection.
+If this callback is not set, the s2n-tls server chooses the first identity in its PSK list that also appears in the client's
+PSK list. In addition to selecting the PSK, **s2n_psk_selection_callback** can be used to set the list of PSKs supported
+by the server prior to the ClientHello.
+
+For Example:
+
+In the following example **s2n_psk_selection_callback** selects a PSK with a shared PSK wired index and sends it to the
+callback via the **context** pointer.
+
+```c
+int s2n_psk_selection_callback(struct s2n_connection *conn, void *context,
+        struct s2n_offered_psk_list *psk_identity_list)
+{
+    DEFER_CLEANUP(struct s2n_offered_psk *offered_psk = s2n_offered_psk_new(), s2n_offered_psk_free);
+    EXPECT_NOT_NULL(offered_psk);
+    uint16_t *wire_index_choice = (uint16_t*) context;
+    uint16_t idx = 0;
+    while(s2n_offered_psk_list_has_next(psk_identity_list)) {
+        POSIX_GUARD(s2n_offered_psk_list_next(psk_identity_list, &offered_psk));
+        if (idx == *wire_index_choice) {
+            POSIX_GUARD(s2n_offered_psk_list_choose_psk(psk_identity_list, &offered_psk));
+            break;
+        }
+        idx++;
+    };
+    return S2N_SUCCESS;
+}
+
+struct s2n_config *config = s2n_config_new();
+uint16_t expected_wire_choice = 0;
+s2n_config_set_psk_selection_callback(config, s2n_psk_selection_callback, &expected_wire_choice));
+```
+
+In the following example **s2n_psk_selection_callback** sets the list of server supported PSKs and also selects
+the PSK to be used for the connection.
+
+```c
+uint8_t identity[] = { 0x01, 0xFF, 0x23 };
+uint8_t secret[] = { 0x0A, 0x03, 0x05, 0x08, 0x01 };
+uint8_t identity_2[] = { 0x56, 0xFA, 0x09, 0x78 };
+uint8_t secret_2[] = { 0x12, 0x30, 0x09, 0x02, 0x04 };
+uint8_t *identity_list[] = { identity, identity_2 };
+uint8_t *secret_list[] = { secret, secret_2 };
+size_t psk_list_len = sizeof(identity_list)/sizeof(*identity_list);
+
+int s2n_setup_psks(struct s2n_connection *conn)
+{
+    for (size_t i = 0; i < psk_list_len; i++) {
+        DEFER_CLEANUP(struct s2n_psk *psk = s2n_external_psk_new(), s2n_psk_free);
+        POSIX_GUARD(s2n_psk_set_identity(psk, identity_list[i], sizeof(identity_list[i])));
+        POSIX_GUARD(s2n_psk_set_secret(psk, secret_list[i], sizeof(secret_list[i])));
+        POSIX_GUARD(s2n_connection_append_psk(conn, psk));
+    }
+    return S2N_SUCCESS;
+}
+
+int s2n_psk_selection_callback(struct s2n_connection *conn, void *context,
+        struct s2n_offered_psk_list *psk_identity_list)
+{
```

Comment:
I worry that using the default logic as your example might confuse customers into thinking they need to re-implement the default logic. Can you use a more likely customer use case, like a server needing to look PSKs up in an external list, possibly while also loading the PSK it needs?